assets/app/scripts/services/data.js

```diff
@@ -504,14 +504,48 @@ angular.module('openshiftConsole')
     }
   };
 
-  DataService.prototype._watchWebsocketRetries = function(type, context, retry) {
+  DataService.prototype._addWatchWebsocketEvent = function(type, context, eventType) {
     var key = this._uniqueKeyForTypeContext(type, context);
-    if (retry === undefined) {
-      return this._watchWebsocketRetriesMap[key];
+    var events = this._watchWebsocketEventsMap[key];
+    if (!events) {
+      events = this._watchWebsocketEventsMap[key] = [];
+    } else if (events.length === 10) {
+      // Only keep 10 events. Shift the array to make room for the new event.
+      events.shift();
+    }
+
+    // Add the event to the end of the array with the time in millis.
+    events.push({
+      type: eventType,
+      time: Date.now()
+    });
+  };
+
+  DataService.prototype._isTooManyWatchWebsocketRetries = function(type, context) {
+    var key = this._uniqueKeyForTypeContext(type, context);
+    var events = this._watchWebsocketEventsMap[key];
+
+    // If there are less than 5 events, always retry.
+    if (!events || events.length < 5) {
```

Comment:
I'd probably skip the early return when `< 5`... that requires magic knowledge about a later condition we're checking for